### PR TITLE
Add deserializing missing properties

### PR DESCRIPTION
**Pull request by kpajdzik**

### CODE REVIEW

File: lib/serializer.ts [hunk 1]
```diff
@@ -554,6 +556,8 @@ function deserializeCompositeType(serializer: Serializer, mapper: CompositeMappe
         if (headerKey.startsWith(headerCollectionPrefix)) {
           dictionary[headerKey.substring(headerCollectionPrefix.length)] = serializer.deserialize((propertyMapper as DictionaryMapper).type.value, responseBody[headerKey], propertyObjectName);
         }
+
+        handledPropertyNames.push(headerKey);
```
Comment (kpajdzik) [inline lib/serializer.ts]: Adding handled properties in every path causes problems with nested properties with conflicting names.

File: lib/serializer.ts [hunk 2]
```diff
@@ -618,6 +622,27 @@ function deserializeCompositeType(serializer: Serializer, mapper: CompositeMappe
         instance[responsePropName] = serializer.deserialize(additionalPropertiesMapper, responseBody[responsePropName], objectName + '["' + responsePropName + '"]');
       }
     }
+  } else {
+    const containsSerializedName = (modelProperties: {[propertyName: string]: Mapper}, serializedKey: string): boolean => {
+      for (const key of Object.keys(modelProperties)) {
```
Comment (sergey-shandar) [inline lib/serializer.ts]: nit: `Object.keys(modelProperties).find(…)`

File: test/shared/serializationTests.ts [hunk 3]
```diff
@@ -1049,11 +1050,12 @@ describe("msrest", function () {
       deserializedSawshark.siblings.length.should.equal(1);
       deserializedSawshark.siblings[0].fishtype.should.equal("mutatedshark");
```
Comment (sergey-shandar) [inline test/shared/serializationTests.ts]: nit: Could we use deep strict equal here to compare two structures?

File: lib/serializer.ts [hunk 4]
```diff
@@ -618,6 +622,27 @@ function deserializeCompositeType(serializer: Serializer, mapper: CompositeMappe
...
+        if (mapper.serializedName === serializedKey) {
+          return true;
+        }
+      }
+
+      return false;
+    };
+
+    const isSpecialProperty = (propertyName: string): boolean => {
+      return [ "$", "_" ].includes(propertyName);
```
Comment (sergey-shandar) [inline lib/serializer.ts]:
```ts
const isSpecialProperty = (propertyName: string): boolean => 
  [ "$", "_"].includes(propertyName)
```

File: lib/serializer.ts [hunk 5]
```diff
@@ -618,6 +622,27 @@ function deserializeCompositeType(serializer: Serializer, mapper: CompositeMappe
         instance[responsePropName] = serializer.deserialize(additionalPropertiesMapper, responseBody[responsePropName], objectName + '["' + responsePropName + '"]');
       }
     }
+  } else {
+    const containsSerializedName = (modelProperties: {[propertyName: string]: Mapper}, serializedKey: string): boolean => {
+      for (const key of Object.keys(modelProperties)) {
+        const mapper: Mapper = modelProperties[key];
```
Comment (sergey-shandar) [inline lib/serializer.ts]: nit:
```ts
for (const [key, mapper] of Object.entries(modelProperties)) { 
   ...
```

File: test/shared/serializationTests.ts (hunk 3 above)
Comment (kpajdzik) [reply test/shared/serializationTests.ts]: Unfortunately, no as there are some properties that are ignored in the comparison.

File: lib/serializer.ts (hunk 2 above)
Comment (daschult) [reply lib/serializer.ts]: I don't agree with this suggestion. I think the foreach pattern is just fine here.

File: lib/serializer.ts (hunk 5 above)
Comment (daschult) [reply lib/serializer.ts]: This is a good suggestion, though.

File: lib/serializer.ts (hunk 4 above)
Comment (daschult) [reply lib/serializer.ts]: This is specific to the XML scenario, so maybe rename this to `isSpecialXMLProperty`. I'd also recommend changing it to use the normal function syntax (`function isSpecialXMLProperty(propertyName: string): boolean =>...`).

File: lib/serializer.ts [hunk 6]
```diff
@@ -532,6 +532,20 @@ function serializeCompositeType(serializer: Serializer, mapper: CompositeMapper,
   return object;
 }
 
+function isSpecialXmlProperty(propertyName: string): boolean {
+  return [ "$", "_" ].includes(propertyName);
+}
+
+function containsSerializedName(modelProperties: {[propertyName: string]: Mapper}, serializedKey: string): boolean {
+  for (const [, mapper] of Object.entries(modelProperties)) {
```
Comment (sergey-shandar) [inline lib/serializer.ts]: nit: :-)
```ts
Object.values(modelProperties).some(mapper => mapper.serializedName === serializedKey)
```

File: lib/serializer.ts [hunk 7]
```diff
@@ -618,6 +631,12 @@ function deserializeCompositeType(serializer: Serializer, mapper: CompositeMappe
         instance[responsePropName] = serializer.deserialize(additionalPropertiesMapper, responseBody[responsePropName], objectName + '["' + responsePropName + '"]');
       }
     }
+  } else {
+    for (const key of Object.keys(responseBody)) {
+      if (instance[key] === undefined && !handledPropertyNames.includes(key) && !containsSerializedName(modelProps, key) && !isSpecialXmlProperty(key)) {
```
Comment (daschult) [inline lib/serializer.ts]: You shouldn't need to check containsSerializedName() anymore, right?

**Pull request merged**